resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala

```diff
@@ -235,35 +263,39 @@ private[spark] class KubernetesSparkRestServer(
       }
     }
 
-    private def writeUploadedJars(files: Option[TarGzippedData], rootTempDir: File):
+    private def writeUploadedJars(jars: TarGzippedData, rootTempDir: File):
         Seq[String] = {
       val resolvedDirectory = new File(rootTempDir, "jars")
       if (!resolvedDirectory.mkdir()) {
         throw new IllegalStateException(s"Failed to create jars dir at " +
           resolvedDirectory.getAbsolutePath)
       }
-      writeBase64ContentsToFiles(files, resolvedDirectory)
+      CompressionUtils.unpackAndWriteCompressedFiles(jars, resolvedDirectory)
     }
 
-    private def writeUploadedFiles(files: Option[TarGzippedData]): Seq[String] = {
+    private def writeUploadedFiles(files: TarGzippedData): Seq[String] = {
       val workingDir = Paths.get("").toFile.getAbsoluteFile
-      writeBase64ContentsToFiles(files, workingDir)
+      CompressionUtils.unpackAndWriteCompressedFiles(files, workingDir)
     }
 
-    def resolvedAppResource(appResource: AppResource, tempDir: File): String = {
-      val appResourcePath = appResource match {
+    // Retrieve the path on the driver container where the main app resource is, and what value it
+    // ought to have in the spark.jars property. The two may be different because for non-local
+    // dependencies, we have to fetch the resource (if it is not "local") but still want to use
+    // the full URI in spark.jars.
+    private def resolvedAppResource(appResource: AppResource, tempDir: File): (String, String) = {
+      appResource match {
         case UploadedAppResource(resourceContentsBase64, resourceName) =>
           val resourceFile = new File(tempDir, resourceName)
           val resourceFilePath = resourceFile.getAbsolutePath
           if (resourceFile.createNewFile()) {
             val resourceContentsBytes = Base64.decodeBase64(resourceContentsBase64)
             Files.write(resourceContentsBytes, resourceFile)
-            resourceFile.getAbsolutePath
+            (resourceFile.getAbsolutePath, resourceFile.getAbsolutePath)
           } else {
             throw new IllegalStateException(s"Failed to write main app resource file" +
               s" to $resourceFilePath")
           }
-        case ContainerAppResource(resource) => resource
+        case ContainerAppResource(resource) => (Utils.resolveURI(resource).getPath, resource)
```

Comment:
Yep this should be fine to not support that driver-only case.  If you can bake something ahead of time into the driver you ought to be able to bake it into the executor image too.  Let's not differentiate between just driver or just executor resources.